consensus/spos/consensusMessageValidator.go

```diff
@@ -0,0 +1,270 @@
+
+import (
+	"bytes"
+	"fmt"
+
+	"github.com/ElrondNetwork/elrond-go-logger"
+	"github.com/ElrondNetwork/elrond-go/consensus"
+	"github.com/ElrondNetwork/elrond-go/core"
+)
+
+func (wrk *Worker) checkConsensusMessageValidity(cnsMsg *consensus.Message) error {
+	if !bytes.Equal(cnsMsg.ChainID, wrk.chainID) {
+		return fmt.Errorf("%w : received chain ID from consensus topic is invalid: %s",
+			ErrInvalidChainID,
+			logger.DisplayByteSlice(cnsMsg.ChainID))
+	}
+
+	err := wrk.checkConsensusMessageValidityForMessageType(cnsMsg)
+	if err != nil {
+		return err
+	}
+
+	msgType := consensus.MessageType(cnsMsg.MsgType)
+
+	isBlockHeaderHashSizeInvalid := wrk.consensusService.IsMessageWithBlockBody(msgType) && cnsMsg.BlockHeaderHash != nil ||
+		!wrk.consensusService.IsMessageWithBlockBody(msgType) && len(cnsMsg.BlockHeaderHash) != core.HashSizeInBytes
+	if isBlockHeaderHashSizeInvalid {
+		return fmt.Errorf("%w : received header hash from consensus topic has an invalid size: %d",
+			ErrInvalidHeaderHashSize,
+			len(cnsMsg.BlockHeaderHash))
+	}
+
+	if len(cnsMsg.PubKey) != core.PublicKeySizeInBytes {
+		return fmt.Errorf("%w : received public key from consensus topic has an invalid size: %d",
+			ErrInvalidPublicKeySize,
+			len(cnsMsg.PubKey))
+	}
+
+	if len(cnsMsg.Signature) != core.SignatureSizeInBytes {
+		return fmt.Errorf("%w : received signature from consensus topic has an invalid size: %d",
+			ErrInvalidSignatureSize,
+			len(cnsMsg.Signature))
+	}
+
+	isNodeInEligibleList := wrk.consensusState.IsNodeInEligibleList(string(cnsMsg.PubKey))
+	if !isNodeInEligibleList {
+		return fmt.Errorf("%w : received message from consensus topic has an invalid public key: %s",
+			ErrNodeIsNotInEligibleList,
+			logger.DisplayByteSlice(cnsMsg.PubKey))
+	}
+
+	if wrk.consensusState.RoundIndex+1 < cnsMsg.RoundIndex {
+		log.Trace("received message from consensus topic has a future round",
+			"msg type", wrk.consensusService.GetStringValue(msgType),
+			"from", cnsMsg.PubKey,
+			"header hash", cnsMsg.BlockHeaderHash,
+			"msg round", cnsMsg.RoundIndex,
+			"round", wrk.consensusState.RoundIndex,
+		)
+
+		return ErrMessageForFutureRound
+	}
+
+	if wrk.consensusState.RoundIndex > cnsMsg.RoundIndex {
+		log.Trace("received message from consensus topic has a past round",
+			"msg type", wrk.consensusService.GetStringValue(msgType),
+			"from", cnsMsg.PubKey,
+			"header hash", cnsMsg.BlockHeaderHash,
+			"msg round", cnsMsg.RoundIndex,
+			"round", wrk.consensusState.RoundIndex,
+		)
+
+		return ErrMessageForPastRound
+	}
+
+	err = wrk.checkSignature(cnsMsg)
+	if err != nil {
+		return fmt.Errorf("%w : verify signature for received message from consensus topic failed: %s",
+			ErrInvalidSignature,
+			err.Error())
+	}
+
+	return nil
+}
+
+func (wrk *Worker) checkConsensusMessageValidityForMessageType(cnsMsg *consensus.Message) error {
+	msgType := consensus.MessageType(cnsMsg.MsgType)
+
+	if wrk.consensusService.IsMessageWithBlockBodyAndHeader(msgType) {
```

Comment:
this could have looked better with a switch on the cnsMsg.MsgType

Reply:
Actually we don't have access to the message type as it is defined in bls package and not in spos. This is why we used consensusService interface for this.